### PR TITLE
rgw: URL-decode S3 and Swift object-copy URLs

### PR DESCRIPTION
**Pull request by mattbenjamin**
Fixes: http://tracker.ceph.com/issues/22121

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4300,27 +4301,26 @@ bool RGWCopyObj::parse_copy_location(const string& url_src, string& bucket_name,
     params_str = url_src.substr(pos + 1);
   }
 
-  std::string dec_src = url_decode(name_str);
-  const char *src = dec_src.c_str();
-
-  if (*src == '/') ++src;
+  boost::string_view dec_src{name_str};
+  if (dec_src[0] == '/') dec_src.remove_prefix(1);
```
Comment (rzarzynski) [inline src/rgw/rgw_op.cc]: I know it isn't new but can we have this `if` broke into two lines?

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-01-16-1725](https://trello.com/c/PBlOeZJA/229-wip-yuri2-testing-2018-01-16-1725)

**Comment by cbodley** [comment]
the teuthology run for this came back with a lot of crashes trying to construct a std::string with a nullptr. here's an [example](http://qa-proxy.ceph.com/teuthology/yuriw-2018-01-16_22:06:10-rgw-wip-yuri2-testing-2018-01-16-2029-distro-basic-smithi/2079401/teuthology.log):
> terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
*** Caught signal (Aborted) **
 in thread 3b605700 thread_name:civetweb-worker
 ceph version 13.0.0-4872-gd1090cb (d1090cb442820801ad70367fd67be83dca35ae98) mimic (dev)
 1: (()+0x1d8310) [0x2e0310]
 2: (()+0xf5e0) [0x5c975e0]
 3: (gsignal()+0x37) [0x10db71f7]
 4: (abort()+0x148) [0x10db88e8]
 5: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x105c2ac5]
 6: (()+0x5ea36) [0x105c0a36]
 7: (()+0x5ea63) [0x105c0a63]
 8: (()+0x5ec83) [0x105c0c83]
 9: (std::__throw_logic_error(char const*)+0x77) [0x106157d7]
 10: (RGWHandler_REST_Obj_S3::op_put()+0x141) [0x4c4291]
 11: (RGWHandler_REST::get_op(RGWRados*)+0x96) [0x479476]
 12: (process_request(RGWRados*, RGWREST*, RGWRequest*, std::string const&, rgw::auth::StrategyRegistry const&, RGWRestfulIO*, OpsLogSocket*, int*)+0xfd7) [0x3e6507]
 13: (RGWCivetWebFrontend::process(mg_connection*)+0x373) [0x298be3]
 14: (()+0x1c927f) [0x2d127f]
 15: (()+0x1caa9e) [0x2d2a9e]
 16: (()+0x7e25) [0x5c8fe25]
 17: (clone()+0x6d) [0x10e7a34d]
2018-01-16 22:27:55.518 3b605700 -1 *** Caught signal (Aborted) **

**Comment by mattbenjamin** [comment]
dehowlerized;  the two linespacing changes mentioned by Radoslaw also made

**Comment by mattbenjamin** [comment]
@yuriw building as wip-rgw-22121 -- will schedule a run

**Pull request merged**